src/core/server/logging/appenders/rolling_file/rolling_file_appender.test.ts

```diff
@@ -0,0 +1,275 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import {
+  createRollingStrategyMock,
+  createTriggeringPolicyMock,
+  LayoutsMock,
+  resetAllMocks,
+  RollingFileContextMock,
+  RollingFileManagerMock,
+} from './rolling_file_appender.test.mocks';
+import { rollingFileAppenderMocks } from './mocks';
+import moment from 'moment-timezone';
+import { LogLevel, LogRecord } from '@kbn/logging';
+import { RollingFileAppender, RollingFileAppenderConfig } from './rolling_file_appender';
+
+const config: RollingFileAppenderConfig = {
+  kind: 'rolling-file',
+  path: '/var/log/kibana.log',
+  layout: {
+    kind: 'pattern',
+    pattern: '%message',
+    highlight: false,
+  },
+  policy: {
+    kind: 'time-interval',
+    interval: moment.duration(4, 'hour'),
+    modulate: true,
+  },
+  strategy: {
+    kind: 'numeric',
+    max: 5,
+    pattern: '-%i',
+  },
+};
+
+const createLogRecord = (parts: Partial<LogRecord> = {}): LogRecord => ({
+  timestamp: new Date(),
+  level: LogLevel.Info,
+  context: 'context',
+  message: 'just a log',
+  pid: 42,
+  ...parts,
+});
+
+const nextTick = () => new Promise((resolve) => setTimeout(resolve, 10));
+
+const createPromiseResolver = () => {
+  let resolve: () => void;
+  let reject: () => void;
+  const promise = new Promise<void>((_resolve, _reject) => {
+    resolve = _resolve;
+    reject = _reject;
+  });
+
+  return {
+    promise,
+    resolve: resolve!,
+    reject: reject!,
+  };
+};
+
+describe('RollingFileAppender', () => {
+  let appender: RollingFileAppender;
+
+  let layout: ReturnType<typeof rollingFileAppenderMocks.createLayout>;
+  let strategy: ReturnType<typeof rollingFileAppenderMocks.createStrategy>;
+  let policy: ReturnType<typeof rollingFileAppenderMocks.createPolicy>;
+  let context: ReturnType<typeof rollingFileAppenderMocks.createContext>;
+  let fileManager: ReturnType<typeof rollingFileAppenderMocks.createFileManager>;
+
+  beforeEach(() => {
+    layout = rollingFileAppenderMocks.createLayout();
+    LayoutsMock.create.mockReturnValue(layout);
+
+    policy = rollingFileAppenderMocks.createPolicy();
+    createTriggeringPolicyMock.mockReturnValue(policy);
+
+    strategy = rollingFileAppenderMocks.createStrategy();
+    createRollingStrategyMock.mockReturnValue(strategy);
+
+    context = rollingFileAppenderMocks.createContext('file-path');
+    RollingFileContextMock.mockImplementation(() => context);
+
+    fileManager = rollingFileAppenderMocks.createFileManager();
+    RollingFileManagerMock.mockImplementation(() => fileManager);
+
+    appender = new RollingFileAppender(config);
+  });
+
+  afterAll(() => {
+    resetAllMocks();
```

Comment:
I'm mocking some class constructors in the test file, so yea.

```ts
// src/core/server/logging/appenders/rolling_file/rolling_file_appender.test.mocks.ts#L42
export const RollingFileManagerMock = jest.fn();
jest.doMock('./rolling_file_manager', () => ({
  RollingFileManager: RollingFileManagerMock,
}));
```

```ts
// src/core/server/logging/appenders/rolling_file/rolling_file_appender.test.ts#L101
context = rollingFileAppenderMocks.createContext('file-path');
RollingFileContextMock.mockImplementation(() => context);
```

I could just `clear` them instead of `reset` as the implementations are redefined during the `beforeEach` block, but as the call was required anyway, `reset` felt better.

Reply:
I'm not sure how jest runs tests. But if it runs them in parallel, the shared state could cause false-positive test failures.